### ISSUE TITLE
HorizontalPodAutoscaler for envoy proxy

### ISSUE THREAD

**Issue by gitanuj** [envoyproxy/gateway#703]
*Description*:
Does envoy gateway have any design to scale out envoy proxy pods? Are there any best practices on how to handle increased number of requests? I'm mostly looking at deploying this to clouds like AWS/GCP/Azure which will allocate application/network Load Balancers but then how can I make sure that envoy proxy pod is not the bottleneck in the gateway and can scale out when needed.

**Comment by danehans** [comment]
HPA is currently not supported by Envoy Gateway. It's a feature that would be beneficial to the project. I have added the feature to the backlog. Let me know if you're interested in working on it and I can assign it to you.

> I'm mostly looking at deploying this to clouds like AWS/GCP/Azure which will allocate application/network Load Balancers...

Envoy Gateway currently creates an ELB when running in AWS.  #648 intends to design an API for managing the Envoy service, PTAL, and comment as needed.

**Comment by gitanuj** [comment]
1. Is there any design doc explaining which pods are involved in the data path for EG? I'm trying to understand if there are any other components other that EnvoyProxy which need to be scaled out as well.
2. Before implementing HPA I wanted to understand if I can simply update the number of replicas [here](https://github.com/envoyproxy/gateway/blob/7c6c37ad72a6fb2403f7ea85b94732c0e14906f7/internal/infrastructure/kubernetes/deployment.go#L131). Is there an API through which this can be exposed?

**Comment by danehans** [comment]
>Is there any design doc explaining which pods are involved in the data path for EG? I'm trying to understand if there are any other components other than EnvoyProxy which need to be scaled out as well.

Only the Envoy proxy deployment needs to be scaled. The [design docs](https://gateway.envoyproxy.io/v0.2.0/design_docs.html) are the best place for this type of info, specifically the system design and config API design docs.

> Before implementing HPA I wanted to understand if I can simply update the number of replicas [here](https://github.com/envoyproxy/gateway/blob/7c6c37ad72a6fb2403f7ea85b94732c0e14906f7/internal/infrastructure/kubernetes/deployment.go#L131). Is there an API through which this can be exposed?

You can update this field but EG will revert it back to the desired state. This field should be configurable using the [EnvoyProxy config API](https://github.com/envoyproxy/gateway/blob/v0.2.0/api/config/v1alpha1/envoyproxy_types.go). If you would like to make this field configurable, please create an issue that describes your use case and xref this issue.

**Comment by arkodg** [comment]
cc @qicz in case you're interesting in adding this to `EnvoyProxy` resource :)

**Comment by qicz** [comment]
> cc @qicz in case you're interested in adding this to `EnvoyProxy` resource :)

I am sorry for the late reply. 

I think we just need to provide the same docs about HPA.  current envoy deployment we can configure more fields and #1398 support affinity etc settings.

**Comment by qicz** [comment]
example 

```yaml
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app.kubernetes.io/component: envoy-demo
    app.kubernetes.io/name: envoy-demo
    app.kubernetes.io/part-of: envoy-demo-ns
  name: envoy-demo
  namespace: envoy-demo-ns
spec:
  progressDeadlineSeconds: 600
  replicas:  1
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app.kubernetes.io/component: envoy-demo
      app.kubernetes.io/name: envoy-demo
      app.kubernetes.io/part-of: envoy-demo-ns
  template:
    metadata:
      labels:
        app.kubernetes.io/component: envoy-demo
        app.kubernetes.io/name: envoy-demo
        app.kubernetes.io/part-of: envoy-demo-ns
    spec:
      serviceAccountName: envoy-demo-sa
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
              - matchExpressions:
                  - key: node-role.kubernetes.io/master
                    operator: In
                    values:
                      - ""
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - podAffinityTerm:
                labelSelector:
                  matchLabels:
                    app.kubernetes.io/component: envoy-demo
                    app.kubernetes.io/name: envoy-demo
                    app.kubernetes.io/part-of: envoy-demo-ns
                namespaces:
                  - envoy-demo-ns
                topologyKey: kubernetes.io/hostname
              weight: 100
      containers:
        - name: envoy-demo
          image: envoy-demo:latest
          imagePullPolicy: IfNotPresent
          resources:
            requests:
              cpu: 10m
              memory: 64Mi
            limits:
              cpu: 1
              memory: 1024Mi
      tolerations:
      - effect: "NoSchedule"
        key: "node-role.kubernetes.io/master"
        operator: "Exists"

---
apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
metadata:
  name: test-hpa
  namespace: envoy-demo-ns
spec:
  maxReplicas: 3
  minReplicas: 1
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment
    name: envoy-demo
  metrics:
    - type: Resource
      resource:
        name: cpu
        target:
          type: Utilization
          averageUtilization: 80
```

**Comment by arkodg** [comment]
thanks for outlining this @qicz, looks like we just need to outline this using docs, and no API changes are needed

**Comment by arkodg** [comment]
@qicz are you planning on working on this one for v0.6.0-rc1 ?

**Comment by ardikabs** [comment]
hi @arkodg @qicz 

is there any update on this matter? I could help to work on this. However, I need some advice on how we should support HPA on Envoy Proxy deployment.

Just for context, I've applied a patch to Envoy Gateway at my organization to make it compatible with HPA when enabled externally. This patch allows EG to determine when the `EnvoyProxy` spec sets the number of replicas to 0, effectively making EG disregard the replicas field in the Envoy Proxy Deployment. However, we think that this implicit solution might not align with the community's goals.

**Comment by arkodg** [comment]
hey @ardikabs assigned this issue to you for now, thanks for volunteering !
I think disabling the replica field by default in the deployment is fine

the question I have is - should EG create the HPA resource or should the end user create one and link it to the generated Envoyproxy deployment

**Comment by ardikabs** [comment]
IMO, it would be better if EG also managed the HPA config with introduce another field such as `hpaSpec` like what Istio did on their gateway through IstioOperator API. So that the existence of the field will override the replicas field on EnvoyProxy API for kubernetes provider. Thoughts @arkodg?

**Comment by arkodg** [comment]
ptal @envoyproxy/gateway-maintainers, please share your thoughts on adding `envoyHpa` field into the `EnvoyProxy` resource

**Comment by zirain** [comment]
keep eye on https://github.com/kubernetes-sigs/gateway-api/discussions/1355

**Comment by gazal-k** [comment]
That part that confuses me is that I see it is possible to control the envoy deployment using the `EnvoyProxy` object, like: https://github.com/envoyproxy/gateway/blob/main/examples/kubernetes/envoy-proxy-config.yaml. However, the helm chart does not seem to do this, instead defining minimal config; https://github.com/envoyproxy/gateway/blob/dcc0ee070489453ce53dfdd4b68d55acc773b34b/charts/gateway-helm/values.tmpl.yaml#L26-L34 and defining the `Deployment` object separately; https://github.com/envoyproxy/gateway/blob/main/charts/gateway-helm/templates/envoy-gateway-deployment.yaml

This makes me think, we can introduce opt-in autoscaling configuration similar to: https://github.com/kubernetes/ingress-nginx/blob/7f723c59855e82614582ff7b2efd1783b1afc2ee/charts/ingress-nginx/values.yaml#L367-L373

**Comment by gazal-k** [comment]
Thanks for that. So, the `Deployment` object in the chart is the controller. The controller then creates another `Deployment` object for the envoy proxy fleet?

**Comment by shawnh2** [comment]
> Thanks for that. So, the `Deployment` object in the chart is the controller. The controller then creates another `Deployment` object for the envoy proxy fleet?

correct

**Comment by arkodg** [comment]
hey @ardikabs are you planning on working on this one ?

**Comment by ardikabs** [comment]
Yes @arkodg, I am still working on it

**Comment by shawnh2** [comment]
closed in favor of #2257
